File: src/lib/manifest/extractors/weapons.ts
import type { WeaponRecord, WeaponPerkColumn, WeaponSlotName, ElementName, Hash } from "../types/records";
import type { Extractor, RawTable, RawTableName } from "../types/services";
import type { RawInventoryItem, RawSocketEntry } from "./rawTypes";
import { asRawEquipmentSlot, asRawDamageType } from "./rawTypes";
import {
  iterItems,
  projectBase,
  toWeaponSlot,
  toAmmoType,
  toElementName,
  findSocketPlug,
  socketPlugHashes,
  isOriginSocket,
  isExcludedPerkSocket,
  getPerkSocketIndexes,
} from "./common";
import { isLegendaryWeaponFramePlug } from "@/lib/synergies/weaponArchetypeSubType";

const ITEM_TYPE_WEAPON = 3;
const TIER_LEGENDARY = 5;
const WEAPON_PERKS_CATEGORY_HASH = 4241085061;

type LoadTable = (table: RawTableName) => Promise<RawTable>;

function isLegendaryWeapon(item: RawInventoryItem): boolean {
  return item.itemType === ITEM_TYPE_WEAPON && item.inventory?.tierType === TIER_LEGENDARY;
}

function buildSlotMap(slotTable: RawTable): Map<number, WeaponSlotName> {
  const map = new Map<number, WeaponSlotName>();
  for (const v of Object.values(slotTable)) {
    const slot = asRawEquipmentSlot(v);
    if (!slot) continue;
    const name = toWeaponSlot(slot.displayProperties.name);
    if (name) map.set(slot.hash, name);
  }
  return map;
}

function buildDamageMap(damageTable: RawTable): Map<number, ElementName> {
  const map = new Map<number, ElementName>();
  for (const v of Object.values(damageTable)) {
    const dt = asRawDamageType(v);
    if (!dt) continue;
    const name = toElementName(dt.displayProperties.name);
    if (name) map.set(dt.hash, name);
  }
  return map;
}

function findFramePlug(
  item: RawInventoryItem,
  itemTable: RawTable,
): RawInventoryItem | undefined {
  return findSocketPlug(item, itemTable, isLegendaryWeaponFramePlug);
}

function collectOriginHashes(
  item: RawInventoryItem,
  itemTable: RawTable,
  plugSets: RawTable,
): Hash[] {
  const hashes: Hash[] = [];
  for (const socket of item.sockets?.socketEntries ?? []) {
    if (!isOriginSocket(socket, itemTable)) continue;
    hashes.push(...socketPlugHashes(socket, plugSets).curated);
  }
  return [...new Set(hashes)];
}

function buildSingleColumn(
  socket: RawSocketEntry,
  plugSets: RawTable,
  colIndex: number,
): WeaponPerkColumn {
  const { curated, randomized } = socketPlugHashes(socket, plugSets);
  return { column: colIndex, curated, randomized };
}

function buildPerkColumns(
  item: RawInventoryItem,
  itemTable: RawTable,
  plugSets: RawTable,
): WeaponPerkColumn[] {
  const indexes = getPerkSocketIndexes(item, WEAPON_PERKS_CATEGORY_HASH);
  const sockets = item.sockets?.socketEntries ?? [];
  const columns: WeaponPerkColumn[] = [];
  let colIdx = 0;

  for (const idx of indexes) {
    const socket = sockets[idx];
    if (!socket) continue;
    if (isExcludedPerkSocket(socket, itemTable)) continue;
    const col = buildSingleColumn(socket, plugSets, colIdx);
    if (col.curated.length === 0 && col.randomized.length === 0) continue;
    columns.push(col);
    colIdx++;
  }

  return columns;
}

async function extractWeapons(loadTable: LoadTable): Promise<WeaponRecord[]> {
  const [itemTable, slotTable, damageTable, plugSets] = await Promise.all([
    loadTable("DestinyInventoryItemDefinition"),
    loadTable("DestinyEquipmentSlotDefinition"),
    loadTable("DestinyDamageTypeDefinition"),
    loadTable("DestinyPlugSetDefinition"),
  ]);

  const slotMap = buildSlotMap(slotTable);
  const damageMap = buildDamageMap(damageTable);
  const result: WeaponRecord[] = [];

  for (const item of iterItems(itemTable)) {
    if (!isLegendaryWeapon(item)) continue;

    const slotHash = item.equippingBlock?.equipmentSlotTypeHash;
    const slot = slotHash != null ? (slotMap.get(slotHash) ?? null) : null;
    if (!slot) continue;

    const dmgHash = item.defaultDamageTypeHash;
    const element = dmgHash != null ? (damageMap.get(dmgHash) ?? null) : null;
    if (!element) continue;

    const ammo = toAmmoType(item.equippingBlock?.ammoType);
    if (!ammo) continue;

    const framePlug = findFramePlug(item, itemTable);
    if (!framePlug) continue;

    result.push({
      ...projectBase(item),
      slot,
      element,
      ammo,
      frame: framePlug.displayProperties.name,
      itemTypeName: item.itemTypeDisplayName ?? "",
      originTraitHashes: collectOriginHashes(item, itemTable, plugSets),
      perkColumns: buildPerkColumns(item, itemTable, plugSets),
    });
  }

  return result;
}

export const weaponsExtractor: Extractor<"weapons"> = {
  store: "weapons",
  extract: (loadTable) => extractWeapons(loadTable),
};

File: flutter/packages/manifest/lib/src/extractors/weapons.dart
import '../raw/raw_types.dart';
import '../types/records.dart';
import '../types/services.dart';
import '../types/stores.dart';
import 'common.dart';

const _itemTypeWeapon = 3;
const _tierLegendary = 5;
const _weaponPerksCategoryHash = 4241085061;

class WeaponsExtractor implements EntityExtractor {
  @override
  MvpStoreName get store => MvpStoreName.weapons;

  @override
  Future<List<Object>> extract(LoadRawTable loadTable) async {
    final itemTable = await loadTable('DestinyInventoryItemDefinition');
    final slotTable = await loadTable('DestinyEquipmentSlotDefinition');
    final damageTable = await loadTable('DestinyDamageTypeDefinition');
    final plugSets = await loadTable('DestinyPlugSetDefinition');

    final slotMap = _buildSlotMap(slotTable);
    final damageMap = _buildDamageMap(damageTable);
    final result = <WeaponRecord>[];

    for (final item in iterItems(itemTable)) {
      if (!_isLegendaryWeapon(item)) continue;

      final slotHash = item.equippingBlock?['equipmentSlotTypeHash'] as num?;
      final slot = slotHash != null ? slotMap[slotHash.toInt()] : null;
      if (slot == null) continue;

      final dmgHash = item.defaultDamageTypeHash;
      final element = dmgHash != null ? damageMap[dmgHash] : null;
      if (element == null) continue;

      final ammo = toAmmoType(
        (item.equippingBlock?['ammoType'] as num?)?.toInt(),
      );
      if (ammo == null) continue;

      final framePlug = findSocketPlug(
        item,
        itemTable,
        isLegendaryWeaponFramePlug,
      );
      if (framePlug == null) continue;

      final base = projectBase(item);
      result.add(
        WeaponRecord(
          hash: base.hash,
          name: base.name,
          searchName: base.searchName,
          icon: base.icon,
          slot: slot,
          element: element,
          ammo: ammo,
          frame: framePlug.displayProperties.name,
          itemTypeName: item.itemTypeDisplayName ?? '',
          originTraitHashes: _collectOriginHashes(item, itemTable, plugSets),
          perkColumns: _buildPerkColumns(item, itemTable, plugSets),
        ),
      );
    }
    return result;
  }
}

bool _isLegendaryWeapon(RawInventoryItem item) {
  final tier = item.inventory?['tierType'] as num?;
  return item.itemType == _itemTypeWeapon && tier?.toInt() == _tierLegendary;
}

Map<int, WeaponSlotName> _buildSlotMap(RawTable slotTable) {
  final map = <int, WeaponSlotName>{};
  for (final v in slotTable.values) {
    final slot = RawEquipmentSlot.tryParse(v);
    if (slot == null) continue;
    final name = toWeaponSlot(slot.displayProperties.name);
    if (name != null) map[slot.hash] = name;
  }
  return map;
}

Map<int, ElementName> _buildDamageMap(RawTable damageTable) {
  final map = <int, ElementName>{};
  for (final v in damageTable.values) {
    final dt = RawDamageType.tryParse(v);
    if (dt == null) continue;
    final name = toElementName(dt.displayProperties.name);
    if (name != null) map[dt.hash] = name;
  }
  return map;
}

List<Hash> _collectOriginHashes(
  RawInventoryItem item,
  RawTable itemTable,
  RawTable plugSets,
) {
  final hashes = <Hash>[];
  for (final socket in item.socketEntries) {
    if (!isOriginSocket(socket, itemTable)) continue;
    hashes.addAll(socketPlugHashes(socket, plugSets).curated);
  }
  return hashes.toSet().toList();
}

List<WeaponPerkColumn> _buildPerkColumns(
  RawInventoryItem item,
  RawTable itemTable,
  RawTable plugSets,
) {
  final indexes = getPerkSocketIndexes(item, _weaponPerksCategoryHash);
  final sockets = item.socketEntries;
  final columns = <WeaponPerkColumn>[];
  var colIdx = 0;

  for (final idx in indexes) {
    if (idx < 0 || idx >= sockets.length) continue;
    final socket = sockets[idx];
    if (isExcludedPerkSocket(socket, itemTable)) continue;
    final plugs = socketPlugHashes(socket, plugSets);
    if (plugs.curated.isEmpty && plugs.randomized.isEmpty) continue;
    columns.add(
      WeaponPerkColumn(
        column: colIdx,
        curated: plugs.curated,
        randomized: plugs.randomized,
      ),
    );
    colIdx++;
  }
  return columns;
}
